File: legacy-node/src/sinceFilter.js
import { isCoordinatorApiRole } from "./orderIntentViews.js";
import {
  getDonorNeighbourhoodSinceQuery,
  getDonorNeighbourhoodWindowMs
} from "./donorNeighbourhoodWindow.js";

export { getDonorNeighbourhoodSinceQuery, getDonorNeighbourhoodWindowMs };

/** Default list window for limited (donor) dashboard — matches photo redaction window. */
export function getDonorListSinceMs() {
  return getDonorNeighbourhoodWindowMs();
}

/** Own-initiation history on mobile / initiator dashboard (no neighbourhood scope). */
export function getInitiatorHistorySinceMs() {
  return 7 * 86_400_000;
}

export function intentActivityMs(record) {
  const raw = record.updated_at || record.created_at;
  const ms = Date.parse(raw || "");
  return Number.isNaN(ms) ? 0 : ms;
}

/**
 * Parse `since` query values such as `1h`, `30m`, `7d`.
 * @returns {number|null} window length in ms, or null if invalid / absent
 */
export function parseSinceQuery(value) {
  if (value == null || value === "") {
    return null;
  }
  const raw = String(value).trim().toLowerCase();
  const match = /^(\d+(?:\.\d+)?)(h|m|d)$/.exec(raw);
  if (!match) {
    return null;
  }
  const amount = Number(match[1]);
  if (!Number.isFinite(amount) || amount <= 0) {
    return null;
  }
  const unitMs = { m: 60_000, h: 3_600_000, d: 86_400_000 };
  return amount * unitMs[match[2]];
}

export function formatSinceQuery(sinceMs) {
  if (sinceMs % 3_600_000 === 0) {
    return `${sinceMs / 3_600_000}h`;
  }
  if (sinceMs % 60_000 === 0) {
    return `${sinceMs / 60_000}m`;
  }
  return `${sinceMs}ms`;
}

/**
 * Donor JWT: neighbourhood feeds cap at the default window (2h); own-history
 * lists without near/locality scope may reach [getInitiatorHistorySinceMs].
 * Coordinators may pass `since` optionally; omit for full history.
 */
export function resolveListSinceMs(role, querySince, options = {}) {
  const parsed = parseSinceQuery(querySince);
  if (isCoordinatorApiRole(role)) {
    return parsed;
  }
  const neighbourhoodActive =
    options.neighbourhoodMode === "near" ||
    options.neighbourhoodMode === "locality";
  const maxMs = neighbourhoodActive
    ? getDonorListSinceMs()
    : getInitiatorHistorySinceMs();
  if (parsed == null) {
    return maxMs;
  }
  return Math.min(parsed, maxMs);
}

export function filterRecordsSince(records, sinceMs, nowMs = Date.now()) {
  if (sinceMs == null || sinceMs <= 0) {
    return records;
  }
  const cutoff = nowMs - sinceMs;
  return records.filter((record) => intentActivityMs(record) >= cutoff);
}

File: src/main/java/org/sharingbridge/integration/geo/SinceFilter.java
package org.sharingbridge.integration.geo;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.sharingbridge.integration.auth.Roles;

public final class SinceFilter {

    private static final Pattern SINCE = Pattern.compile("^(\\d+(?:\\.\\d+)?)(h|m|d)$");
    private static final long INITIATOR_HISTORY_MS = 7L * 86_400_000L;

    private SinceFilter() {}

    public static long getDonorListSinceMs() {
        return DonorNeighbourhoodWindow.getDonorNeighbourhoodWindowMs();
    }

    public static long getInitiatorHistorySinceMs() {
        return INITIATOR_HISTORY_MS;
    }

    public static long intentActivityMs(Map<String, Object> record) {
        Object raw = record.get("updated_at");
        if (raw == null) {
            raw = record.get("created_at");
        }
        return parseIsoMs(raw);
    }

    public static Long parseSinceQuery(Object value) {
        if (value == null || String.valueOf(value).isEmpty()) {
            return null;
        }
        String raw = String.valueOf(value).trim().toLowerCase();
        Matcher match = SINCE.matcher(raw);
        if (!match.matches()) {
            return null;
        }
        try {
            double amount = Double.parseDouble(match.group(1));
            if (!Double.isFinite(amount) || amount <= 0) {
                return null;
            }
            long unitMs = switch (match.group(2)) {
                case "m" -> 60_000L;
                case "h" -> 3_600_000L;
                case "d" -> 86_400_000L;
                default -> 0L;
            };
            return (long) (amount * unitMs);
        } catch (NumberFormatException ex) {
            return null;
        }
    }

    public static String formatSinceQuery(long sinceMs) {
        if (sinceMs % 3_600_000L == 0) {
            return (sinceMs / 3_600_000L) + "h";
        }
        if (sinceMs % 60_000L == 0) {
            return (sinceMs / 60_000L) + "m";
        }
        return sinceMs + "ms";
    }

    public static Long resolveListSinceMs(String role, String querySince, String neighbourhoodMode) {
        Long parsed = parseSinceQuery(querySince);
        if (Roles.isCoordinatorApiRole(role)) {
            return parsed;
        }
        boolean neighbourhoodActive =
                "near".equals(neighbourhoodMode) || "locality".equals(neighbourhoodMode);
        long maxMs = neighbourhoodActive ? getDonorListSinceMs() : getInitiatorHistorySinceMs();
        if (parsed == null) {
            return maxMs;
        }
        return Math.min(parsed, maxMs);
    }

    public static List<Map<String, Object>> filterRecordsSince(
            List<Map<String, Object>> records, Long sinceMs, long nowMs) {
        if (sinceMs == null || sinceMs <= 0) {
            return records;
        }
        long cutoff = nowMs - sinceMs;
        List<Map<String, Object>> out = new ArrayList<>();
        for (Map<String, Object> record : records) {
            if (intentActivityMs(record) >= cutoff) {
                out.add(record);
            }
        }
        return out;
    }

    public static long parseIsoMs(Object raw) {
        if (raw == null) {
            return 0;
        }
        try {
            return Instant.parse(String.valueOf(raw)).toEpochMilli();
        } catch (RuntimeException ex) {
            return 0;
        }
    }
}
